src/main/scala/org/constellation/session/SessionTokenService.scala

```diff
@@ -0,0 +1,86 @@
+package org.constellation.session
+
+import cats.effect.Concurrent
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import io.chrisdavenport.fuuid.FUUID
+import io.circe.generic.semiauto._
+import io.circe.{Decoder, Encoder}
+import org.constellation.session.SessionTokenService._
+import org.http4s.util.CaseInsensitiveString
+
+class SessionTokenService[F[_]: Concurrent] {
+
+  private[session] val peersTokens: Ref[F, Map[String, Token]] = Ref.unsafe(Map.empty)
+
+  private[session] val ownToken: Ref[F, Option[Token]] = Ref.unsafe(None)
+
+  def createAndSetNewOwnToken: F[Token] =
+    for {
+      token <- createToken
+      _ <- ownToken.modify(_ => (token.some, ()))
+    } yield token
+
+  def getOwnToken: F[Option[Token]] =
+    ownToken.get
+
+  def updatePeerToken(ip: String, token: Token): F[Unit] =
+    peersTokens.modify(tokens => (tokens + (ip -> token), ()))
+
+  def getPeerToken(ip: String): F[Option[Token]] =
+    peersTokens.get.map(_.get(ip))
+
+  def verifyPeerToken(ip: String, oHeaderToken: Option[Token]): F[TokenVerificationResult] =
+    for {
+      oPeerToken <- getPeerToken(ip)
+      verificationResult = SessionTokenService.compareTokens(oPeerToken, oHeaderToken)
+    } yield verificationResult
+
+  def clearOwnToken: F[Unit] =
+    ownToken.modify(_ => (None, ()))
+
+  def clearPeerToken(ip: String): F[Unit] =
+    peersTokens.modify(tokens => (tokens - ip, ()))
+
+  def clear: F[Unit] =
+    for {
+      _ <- clearOwnToken
+      _ <- peersTokens.modify(_ => (Map.empty, ()))
+    } yield ()
+
+  private[session] def createToken: F[Token] =
+    for {
+      uuid <- FUUID.randomFUUID
+      token = new Token(uuid.toString)
+    } yield token
+
+}
+
+object SessionTokenService {
+  class Token(val value: String) extends AnyVal
+
+  object Token {
+    implicit val tokenEncoder: Encoder[Token] = deriveEncoder
+    implicit val tokenDecoder: Decoder[Token] = deriveDecoder
+  }
+
+  val XSessionToken = CaseInsensitiveString("X-Session-Token")
+
+  trait TokenVerificationResult
+  case object EmptyPeerToken extends TokenVerificationResult
+  case object EmptyHeaderToken extends TokenVerificationResult
+  case object TokensDontMatch extends TokenVerificationResult
+  case object TokenValid extends TokenVerificationResult
+
+  def compareTokens(peerToken: Option[Token], headerToken: Option[Token]): TokenVerificationResult =
+    (peerToken, headerToken) match {
+      case (Some(pToken), Some(hToken)) if pToken == hToken =>
```

Comment:
you can compare option types as long as you have case class not class inside.
just put (None, _) and (_, None) before

Reply:
Isn't Token which is extending AnyVal compared by value, so String?

Reply:
So you created `class Token` and forced comparing by value by using `extends AnyVal`?
Why not just `case class Token`?

Reply:
https://docs.scala-lang.org/overviews/core/value-classes.html

From docs:

```
class Wrapper(val underlying: Int) extends AnyVal
```
>It has a single, public val parameter that is the underlying runtime representation. The type at compile time is Wrapper, but at runtime, the representation is an Int


So in our example @marcinwadon during compile time the representation of `Token` is `class Token(val value: String)` but during runtime representation is just `String` which is compared by value. It is fine imo as it will be just:

```scala
case (Some(String), Some(String)) =>
```

Reply:
From the same doc:
`case class P(val i: Int) extends AnyVal` which gives you setoid instance and apply method not to use new Token(uuid).